crates/matrix-sdk-crypto/src/machine.rs

```diff
@@ -1552,15 +1548,86 @@ impl OlmMachine {
+        };
+        if !device.is_cross_signed_by_owner() {
+            return Err(MegolmError::SenderCrossSigningIdentityUnknown);
+        }
+        if device.is_cross_signing_trusted() {
+            return Ok(());
+        }
+        if sender == self.inner.user_id {
+            // if we get here, the device (that claims to be ours) wasn't
+            // cross-signed by us, so we reject it
+            Err(MegolmError::SenderCrossSigningIdentityUnknown)
+        } else if require_verified {
+            Err(MegolmError::SenderCrossSigningUntrusted)
+        } else {
+            // we don't require the sender to be trusted, but we require that
+            // the sender's identity matches their pinned identity
+            let identity =
+                device.device_owner_identity.expect("Cross-signed device must have owner identity");
+            if identity.other().expect("Device is not our own device").has_pin_violation() {
+                Err(MegolmError::SenderCrossSigningIdentityChanged)
+            } else {
+                Ok(())
+            }
+        }
+    }
+
+    /// Check that the sender of a Megolm session satisfies the trust
+    /// requirement from the decryption settings.
+    async fn check_sender_trust_requirement(
+        &self,
+        session: &InboundGroupSession,
+        sender: &UserId,
+        decryption_settings: &DecryptionSettings,
+    ) -> MegolmResult<()> {
+        match decryption_settings.trust_requirement {
+            TrustRequirement::Untrusted => Ok(()),
+            TrustRequirement::CrossSignedOrLegacy => match session.sender_data {
+                SenderData::SenderKnown { .. } => Ok(()),
+                SenderData::DeviceInfo { legacy_session: true, .. } => Ok(()),
+                SenderData::UnknownDevice { legacy_session: true, .. } => Ok(()),
+                _ => self.check_sender_trusted(session.sender_key(), sender, false).await,
+            },
+            TrustRequirement::CrossSigned => match session.sender_data {
+                SenderData::SenderKnown { .. } => Ok(()),
+                _ => self.check_sender_trusted(session.sender_key(), sender, false).await,
+            },
+            TrustRequirement::VerifiedUser => match session.sender_data {
+                SenderData::SenderKnown { master_key_verified: true, .. } => Ok(()),
+                _ => self.check_sender_trusted(session.sender_key(), sender, true).await,
+            },
+        }
+    }
+
     async fn decrypt_megolm_events(
         &self,
         room_id: &RoomId,
         event: &EncryptedEvent,
         content: &SupportedEventEncryptionSchemes<'_>,
+        decryption_settings: &DecryptionSettings,
     ) -> MegolmResult<(JsonObject, EncryptionInfo)> {
         let session =
             self.get_inbound_group_session_or_error(room_id, content.session_id()).await?;
 
+        self.check_sender_trust_requirement(&session, &event.sender, decryption_settings).await?;
```

Comment:
Shouldn't this use the `get_encryption_info()` method or be incorporated into it? This way we're now fetching the device twice, once for this method and another time for the `get_encryption_info()` method.

It's also fetching the `sender_data` twice, the `get_encryption_info()` method fetches and updates this piece of information as well, sure it's turning it into a `VerificationState` in the end but we can modify the place where we do the conversion.

Reply:
Done